File: backend/app/services/faculty_service.py
from sqlalchemy.orm import Session
from uuid import UUID
from typing import List, Optional
import base64
from models.faculty import Faculty
from models.user import User
from schemas.requests.faculty import FacultyCreate, FacultyUpdate, FacultyImageUpdate
from schemas.responses.faculty import FacultyResponse

def get_all_faculty(db: Session) -> List[FacultyResponse]:
    faculties = db.query(Faculty).all()
    result = []
    
    for faculty in faculties:
        faculty_response = get_faculty_by_id(db, faculty.id)
        if faculty_response:
            result.append(faculty_response)
            
    return result

def get_faculty_by_id(db: Session, faculty_id: UUID) -> Optional[FacultyResponse]:
    faculty = db.query(Faculty).filter(Faculty.id == faculty_id).first()
    if not faculty:
        return None
    user = db.query(User).filter(User.id == faculty.user_id).first()
    image_b64 = base64.b64encode(user.image).decode('utf-8') if user and user.image else None
    return FacultyResponse(
        id=faculty.id,
        user_id=faculty.user_id,
        office_room_id=getattr(faculty, 'office_room_id', None),
        full_name=getattr(faculty, 'full_name', None),
        email=getattr(faculty, 'email', None),
        phone_number=getattr(faculty, 'phone_number', None),
        specialization=getattr(faculty, 'specialization', None),
        research_areas=getattr(faculty, 'research_areas', None),
        employment_status=getattr(faculty, 'employment_status', None),
        designation=getattr(faculty, 'designation', None),
        department=getattr(faculty, 'department', None),
        experience=getattr(faculty, 'experience', None),
        number_of_publications=getattr(faculty, 'number_of_publications', None),
        qualifications=getattr(faculty, 'qualifications', None),
        image=image_b64
    )

def create_faculty(db: Session, data: FacultyCreate) -> FacultyResponse:
    faculty_data = data.dict(exclude={"image"})
    faculty = Faculty(**faculty_data)
    db.add(faculty)
    db.flush()
    # Set user image if provided
    if getattr(data, "image", None):
        user = db.query(User).filter(User.id == data.user_id).first()
        if user:
            user.image = base64.b64decode(data.image)
    db.commit()
    db.refresh(faculty)
    return get_faculty_by_id(db, faculty.id)

def update_faculty(db: Session, faculty_id: UUID, data: FacultyUpdate) -> Optional[FacultyResponse]:
    faculty = db.query(Faculty).filter(Faculty.id == faculty_id).first()
    if not faculty:
        return None
    update_data = data.dict(exclude_unset=True, exclude={"image"})
    for key, value in update_data.items():
        setattr(faculty, key, value)
    # Update image if provided
    if getattr(data, "image", None):
        user = db.query(User).filter(User.id == faculty.user_id).first()
        if user:
            user.image = base64.b64decode(data.image)
    db.commit()
    db.refresh(faculty)
    return get_faculty_by_id(db, faculty.id)

def update_faculty_image(db: Session, faculty_id: UUID, image_in: FacultyImageUpdate) -> Optional[FacultyResponse]:
    faculty = db.query(Faculty).filter(Faculty.id == faculty_id).first()
    if not faculty:
        return None
    user = db.query(User).filter(User.id == faculty.user_id).first()
    if not user:
        return None
    user.image = base64.b64decode(image_in.image)
    db.commit()
    return get_faculty_by_id(db, faculty.id)

def delete_faculty(db: Session, faculty_id: UUID) -> bool:
    faculty = db.query(Faculty).filter(Faculty.id == faculty_id).first()
    if not faculty:
        return False
    db.delete(faculty)
    db.commit()
    return True

File: Frontend/src/api/FacultyService.js
/**
 * Faculty service for handling faculty data operations
 * This service handles CRUD operations for faculty members
 */
import ApiService from './ApiService';
import API_CONFIG from './config';
import authService from './AuthService'; 

class FacultyService {
  constructor() {
    // Ensure this is a singleton
    if (FacultyService.instance) {
      return FacultyService.instance;
    }
    FacultyService.instance = this;
    
    // Tahsin added - Define faculty endpoints
    this.endpoints = {
      BASE: '/faculty',
      GET_ALL: '/faculty',
      GET_BY_ID: (id) => `/faculty/${id}`,
      CREATE: '/faculty',
      UPDATE: (id) => `/faculty/${id}`,
      DELETE: (id) => `/faculty/${id}`,
    };
  }

  /**
   * Get all faculty members with optional pagination
   * @param {number} skip - Number of records to skip
   * @param {number} limit - Maximum number of records to return
   * @returns {Promise} - Promise with faculty data
   */
  async getAllFaculty(skip = 0, limit = 100) {
    try {
      const response = await ApiService.get(this.endpoints.GET_ALL, { skip, limit });
      return response;
    } catch (error) {
      console.error('Failed to get faculty data:', error);
      throw error;
    }
  }

  /**
   * Get faculty member by ID
   * @param {number} id - Faculty ID
   * @returns {Promise} - Promise with faculty data
   */
  async getFacultyById(id) {
    try {
      const response = await ApiService.get(this.endpoints.GET_BY_ID(id));
      return response;
    } catch (error) {
      console.error(`Failed to get faculty with ID ${id}:`, error);
      throw error;
    }
  }

  /**
   * Create a new faculty member with multipart/form-data
   * @param {FormData} formData - FormData object containing faculty information and profile photo
   * @returns {Promise} - Promise with created faculty data
   * 
   * Tahsin added - This method handles multipart/form-data submission for faculty creation
   * including profile photo upload
   */

  async createFaculty(faculty) {
    try {
    // Step 1: Sign up the user
      const userPayload = {
        user_name: faculty.user_name,
        email: faculty.email,
        password: faculty.password,
        full_name: faculty.name,
        role: "faculty"
      };

      const userResponse = await authService.signup(userPayload);

      if (!userResponse || !userResponse.id) {
        throw new Error("User signup failed — no user ID returned.");
      }

      const userId = userResponse.id;

      console.log("User created with ID:", userId);

      // Step 2: Build JSON object for FacultyCreate
      const facultyPayload = {
        user_id: userId,
        office_room_id: faculty.office_room_id || null,
        full_name: faculty.name,
        email: faculty.email,
        phone_number: faculty.phone || null,
        specialization: faculty.specialization || null,
        research_areas: faculty.researchAreas || null,
        employment_status: "Active",
        designation: faculty.designation,
        department: "CSE",
        experience: faculty.experience || null,
        number_of_publications: parseInt(faculty.publications) || 0,
        qualifications: faculty.qualifications || null,
        image: null
      };

      console.log("Faculty payload:", facultyPayload);

      // Handle optional profile image (convert to base64 if it's a data URL)
      if (faculty.profileImage?.startsWith("data:image/")) {
        const base64 = faculty.profileImage.split(",")[1];
        facultyPayload.image = base64;
      }

      // Step 3: Call the faculty creation API
      const token = localStorage.getItem("token");
      const response = await fetch("http://localhost:8000/api/v1/faculty/", {
        method: "POST",
        body: JSON.stringify(facultyPayload)
      });

      if (!response.ok) {
        const errorData = await response.json();
        throw new Error(`Faculty creation failed: ${errorData.message || response.statusText}`);
      }

      alert("Faculty created successfully!");
      fetchAllFaculty(); // Refresh list

    } catch (err) {
      console.error("Failed to create faculty:", err);
      alert("Failed to create faculty. Check console for details.");
    }
  }

  /**
   * Update an existing faculty member with multipart/form-data
   * @param {number} id - Faculty ID
   * @param {FormData} formData - FormData object containing faculty information and profile photo
   * @returns {Promise} - Promise with updated faculty data
   * 
   * Tahsin added - This method handles multipart/form-data submission for faculty updates
   * including profile photo upload
   */
  async updateFaculty(id, formData) {
    try {
      // Tahsin added - Special handling for multipart/form-data
      const url = `${API_CONFIG.BASE_URL}${API_CONFIG.API_VERSION}${this.endpoints.UPDATE(id)}`;
      
      // Get authentication token if available
      const token = localStorage.getItem('token');
      const headers = {};
      
      if (token) {
        headers['Authorization'] = `Bearer ${token}`;
      }
      
      // Make direct fetch call for multipart/form-data
      const response = await fetch(url, {
        method: 'PUT',
        headers,
        body: formData,
        mode: 'cors'
      });
      
      // Use ApiService's response handler for consistent error handling
      return await ApiService.handleResponse(response);
    } catch (error) {
      console.error(`Failed to update faculty with ID ${id}:`, error);
      throw error;
    }
  }

  /**
   * Delete a faculty member
   * @param {number} id - Faculty ID
   * @returns {Promise} - Promise with deletion status
   */
  async deleteFaculty(id) {
    try {
      const response = await ApiService.delete(this.endpoints.DELETE(id));
      return response;
    } catch (error) {
      console.error(`Failed to delete faculty with ID ${id}:`, error);
      throw error;
    }
  }
}

// Create and export a singleton instance
const facultyService = new FacultyService();
export default facultyService;
